### PR TITLE
refactor(sdk)!: Stateless `Stream` 

### PR DESCRIPTION
**Pull request by teogeb**
**This is a breaking changes as this changes the API** 

`Stream` instances no longer have metadata field. Instead the metadata is queried from `StreamRegistry`. 
- `StreamRegistry` already caches the metadata. Now we explicity populate the cache e.g. in `searchStreams()` (which queries `id` + `metadata`) from The Graph).

## API

These `Stream` methods are now async:
- `getMetadata()`
- `getPartitionCount()`
- `getDescription()`
- `getStorageDayCount()`

Also added `StreamrClient#getStreamMetadata()` method which is used by `Stream#getMetadata()`.

## Testing

Added new end-to-end tests to assert cache population logic. The new `ProxyHttpServer` helper class which is needed by those tests.

### CODE REVIEW

File: packages/node/src/plugins/storage/StorageConfig.ts [hunk 1]
```diff
@@ -51,14 +51,14 @@ export class StorageConfig {
         this.clusterSize = clusterSize
         this.myIndexInCluster = myIndexInCluster
         this.listener = listener
-        this.storagePoller = new StoragePoller(clusterId, pollInterval, streamrClient, (streams, block) => {
-            const streamParts = streams.flatMap((stream: Stream) => ([
-                ...this.createMyStreamParts(stream)
-            ]))
+        this.storagePoller = new StoragePoller(clusterId, pollInterval, streamrClient, async (streams, block) => {
```
Comment (harbu) [inline packages/node/src/plugins/storage/StorageConfig.ts:54]: Potential unhandled promise rejection here, does the `StoragePoller` handle async callback and does it handle the error ?

File: packages/node/src/plugins/storage/StorageConfig.ts [hunk 2]
```diff
@@ -51,14 +51,14 @@ export class StorageConfig {
...
+            const streamParts = (await Promise.all(streams.map(async (stream: Stream) => {
+                return [...await this.createMyStreamParts(stream)]
+            }))).flat()
             this.handleDiff(this.synchronizer.ingestSnapshot(new Set<StreamPartID>(streamParts), block))
         })
-        this.storageEventListener = new StorageEventListener(clusterId, streamrClient, (stream, type, block) => {
-            const streamParts = this.createMyStreamParts(stream)
+        this.storageEventListener = new StorageEventListener(clusterId, streamrClient, async (stream, type, block) => {
```
Comment (harbu) [inline packages/node/src/plugins/storage/StorageConfig.ts:60]: Potential unhandled promise rejection here, does the `StorageEventListener` handle async callback and does it handle the error ?

File: packages/node/src/plugins/storage/StorageEventListener.ts [hunk 3]
```diff
@@ -11,8 +11,8 @@ export class StorageEventListener {
     private readonly clusterId: EthereumAddress
     private readonly streamrClient: StreamrClient
     private readonly onEvent: (stream: Stream, type: 'added' | 'removed', block: number) => void
-    private readonly onAddToStorageNode: (event: StorageNodeAssignmentEvent) => void
-    private readonly onRemoveFromStorageNode: (event: StorageNodeAssignmentEvent) => void
+    private readonly onAddToStorageNode: (event: StorageNodeAssignmentEvent) => unknown
+    private readonly onRemoveFromStorageNode: (event: StorageNodeAssignmentEvent) => unknown
```
Comment (harbu) [inline packages/node/src/plugins/storage/StorageEventListener.ts]: Why are these changes needed?

File: packages/node/src/plugins/storage/StoragePoller.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class StoragePoller {
         clusterId: string,
         pollInterval: number,
         streamrClient: StreamrClient,
-        onNewSnapshot: (streams: Stream[], block: number) => void
+        onNewSnapshot: (streams: Stream[], block: number) => unknown
```
Comment (harbu) [inline packages/node/src/plugins/storage/StoragePoller.ts]: Why needed?

File: packages/sdk/src/StreamrClient.ts [hunk 5]
```diff
@@ -426,12 +440,16 @@ export class StreamrClient {
         permissionFilter: SearchStreamsPermissionFilter | undefined,
         orderBy: SearchStreamsOrderBy = { field: 'id', direction: 'asc' }
     ): AsyncIterable<Stream> {
-        return this.streamRegistry.searchStreams(
+        logger.debug('Search for streams', { term, permissionFilter })
+        if ((term === undefined) && (permissionFilter === undefined)) {
+            throw new Error('Requires a search term or a permission filter')
+        }
```
Comment (harbu) [inline packages/sdk/src/StreamrClient.ts:446]: Is this a new limitation?

File: packages/sdk/src/StreamrClient.ts [hunk 6]
```diff
@@ -564,8 +582,11 @@ export class StreamrClient {
      */
     async getStoredStreams(storageNodeAddress: HexString): Promise<{ streams: Stream[], blockNumber: number }> {
         const queryResult = await this.streamStorageRegistry.getStoredStreams(toEthereumAddress(storageNodeAddress))
+        for (const stream of queryResult.streams) {
+            this.streamRegistry.populateMetadataCache(stream.id, stream.metadata)
+        }
```
Comment (harbu) [inline packages/sdk/src/StreamrClient.ts:587]: Let's discuss the implications of filling cache based on Graph.

File: packages/sdk/src/contracts/StreamRegistry.ts [hunk 7]
```diff
@@ -295,19 +293,21 @@ export class StreamRegistry {
...
             term,
             permissionFilter,
             orderBy,
-            this.theGraphClient,
-            this.logger,
-            client)
+            this.theGraphClient)
+        for await (const item of queryResult) {
+            const id = toStreamID(item.stream.id)
+            this.populateMetadataCache(id, parseMetadata(item.stream.metadata))
```
Comment (harbu) [inline packages/sdk/src/contracts/StreamRegistry.ts:308]: Let's discuss the implications of filling cache based on Graph.

File: packages/sdk/test/end-to-end/contract-call-cache.test.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            const stream = (await nextValue(client.searchStreams(existingStreamId, undefined)[Symbol.asyncIterator]()))!
+            await stream.getMetadata()
+            expect(getMethodCalls()).toHaveLength(0)
+        })
+
+        it('cache updated when calling setStreamMetatadata()', async () => {
+            const NEW_METADATA = { foo: Date.now() }
+            await client.setStreamMetadata(existingStreamId, NEW_METADATA)
+            expect(await client.getStreamMetadata(existingStreamId)).toEqual(NEW_METADATA)
+            expect(getMethodCalls()).toHaveLength(0)
```
Comment (harbu) [inline packages/sdk/test/end-to-end/contract-call-cache.test.ts:97]: Should this be?
```suggestion
            expect(getMethodCalls()).toHaveLength(1)
```

File: packages/sdk/test/end-to-end/contract-call-cache.test.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            creator.grantPermissions(existingStreamId, {
+                userId: authenticatedUser.address,
+                permissions: [StreamPermission.EDIT]
+            })
+            await creator.destroy()
+            await waitForTheGraphToHaveIndexed(existingStreamId, creator)
+        })
+
+        beforeEach(async () => {
+            server = new ProxyHttpServer(CHAIN_CONFIG.dev2.rpcEndpoints[0].url)
```
Comment (harbu) [inline packages/sdk/test/end-to-end/contract-call-cache.test.ts:54]: FWIW we could also test this same behavior without an actual server if we were able to somehow mock the `streamRegistryContractReadonly` instance within `StreamRegistry`.

File: packages/node/src/plugins/storage/StorageEventListener.ts (hunk 3 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StorageEventListener.ts]: So that the implementation can be async method (can return e.g. `Promise<void>`). Alternatively we could use `setImmedate`/`executeSafePromise` wrapper in the `onAddToStorageNode` implementation.

File: packages/node/src/plugins/storage/StoragePoller.ts (hunk 4 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StoragePoller.ts]: Similar to https://github.com/streamr-dev/network/pull/2883#discussion_r1850472346

File: packages/sdk/src/StreamrClient.ts (hunk 5 above)
Comment (teogeb) [reply packages/sdk/src/StreamrClient.ts:446]: No. Moved from `searchStream.ts#47` (this is the only caller). Good to validate user input in the API method, i.e. in the `StreamrClient` instead of `searchStreams` utility.

File: packages/sdk/test/end-to-end/contract-call-cache.test.ts (hunk 8 above)
Comment (teogeb) [reply packages/sdk/test/end-to-end/contract-call-cache.test.ts:97]: No. The `setStreamMetadata()` populates the cache, and this test that the cache value is used. I.e. no call to contract is needed.

File: packages/sdk/test/end-to-end/contract-call-cache.test.ts (hunk 9 above)
Comment (teogeb) [reply packages/sdk/test/end-to-end/contract-call-cache.test.ts:54]: Yes, then it would be unit test.

File: packages/node/src/plugins/storage/StorageConfig.ts (hunk 2 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StorageConfig.ts:60]: Fixed

File: packages/node/src/plugins/storage/StoragePoller.ts (hunk 4 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StoragePoller.ts]: Changed in commit bf9df6681

File: packages/node/src/plugins/storage/StorageEventListener.ts (hunk 3 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StorageEventListener.ts]: Reverted in 20ab88664

File: packages/node/src/plugins/storage/StorageConfig.ts (hunk 1 above)
Comment (teogeb) [reply packages/node/src/plugins/storage/StorageConfig.ts:54]: `StoragePoller#tryPoll` catches rejection (it calls `this.poll()`, and `poll()` calls this callback)

File: packages/sdk/src/StreamrClient.ts (hunk 6 above)
Comment (harbu) [reply packages/sdk/src/StreamrClient.ts:587]: This is fine based on discussion

File: packages/sdk/src/contracts/StreamRegistry.ts (hunk 7 above)
Comment (harbu) [reply packages/sdk/src/contracts/StreamRegistry.ts:308]: This is fine based on discussion

**Review by harbu — APPROVED** [review]
